<h2>Description</h2>
<p>This track shows the annotation.</p>
<p>The Id of each track can be interpreted by splitting the string on each pipe
symbol "|". The fields are:</p>
<ol>
	<li>Annotation type</li>
	<li>Conitg/ Plasmid Id</li>
	<li>Feature type</li>
	<li>Annotation Id</li>
</ol>
<h3>Annotation Details</h3>
<table class="bedExtraTbl">
<tr>
<td>Annotation Provider</td>
<td>NCBI</td>
</tr>
<tr>
<td>Annotation Date</td>
<td>04/10/2019 15:59:01</td>
</tr>
<tr>
<td>Annotation Pipeline</td>
<td>NCBI Prokaryotic Genome Annotation Pipeline</td>
</tr>
<tr>
<td>Annotation Method</td>
<td>Best-placed reference protein set; GeneMarkS-2+</td>
</tr>
<tr>
<td>Annotation Software revision</td>
<td>4.8</td>
</tr>
<tr>
<td>Features Annotated</td>
<td>Gene; CDS; rRNA; tRNA; ncRNA; repeat_region</td>
</tr>
<tr>
<td>Genes (total)</td>
<td>5,766</td>
</tr>
<tr>
<td>CDSs (total)</td>
<td>5,678</td>
</tr>
<tr>
<td>Genes (coding)</td>
<td>5,481</td>
</tr>
<tr>
<td>CDSs (with protein)</td>
<td>5,481</td>
</tr>
<tr>
<td>Genes (RNA)</td>
<td>88</td>
</tr>
<tr>
<td>rRNAs</td>
<td>5, 5, 5 (5S, 16S, 23S)</td>
</tr>
<tr>
<td>complete rRNAs</td>
<td>5, 5, 5 (5S, 16S, 23S)</td>
</tr>
<tr>
<td>tRNAs</td>
<td>69</td>
</tr>
<tr>
<td>ncRNAs</td>
<td>4</td>
</tr>
<tr>
<td>Pseudo Genes (total)</td>
<td>197</td>
</tr>
<tr>
<td>CDSs (without protein)</td>
<td>197</td>
</tr>
<tr>
<td>Pseudo Genes (ambiguous residues)</td>
<td>0 of 197</td>
</tr>
<tr>
<td>Pseudo Genes (frameshifted)</td>
<td>87 of 197</td>
</tr>
<tr>
<td>Pseudo Genes (incomplete)</td>
<td>97 of 197</td>
</tr>
<tr>
<td>Pseudo Genes (internal stop)</td>
<td>48 of 197</td>
</tr>
<tr>
<td>Pseudo Genes (multiple problems)</td>
<td>28 of 197</td>
</tr>
<tr>
<td>CRISPR Arrays</td>
<td>6</td>
</tr>
</table>

<h2>Display Conventions and Configuration</h2>
<p>The colours of the annotation corresponds with the reading frame of the gene:</p>
<ul>
<li><span style='border: 2px inset #000000; background-color:#ff1a1a;'>&nbsp;Red&nbsp;</span> + strand 1 frame
<li><span style='border: 2px inset #000000; background-color:#1a1aff;'>&nbsp;Blue&nbsp;</span> + strand 2 frame
<li><span style='border: 2px inset #000000; background-color:#2db92d;'>&nbsp;Green&nbsp;</span> + strand 3 frame
<li><span style='border: 2px inset #000000; background-color:#ac00e6;'>&nbsp;Purple&nbsp;</span> - strand 4 frame
<li><span style='border: 2px inset #000000; background-color:#e67300;'>&nbsp;Orange&nbsp;</span> - strand 5 frame
<li><span style='border: 2px inset #000000; background-color:#00e6b8;'>&nbsp;Turquoise&nbsp;</span> - strand 6 frame
</ul>

<h2>Credits</h2>

<p>
Data were processed at the Bioinformatic Institute of Leipzig. For inquiries, please
contact us at the following address: john (at) bioinf.uni-leipzig.de
</p>

<h2>References</h2>

<h3>DBLINK</h3>
<table class="bedExtraTbl">
<tr>
<td>BioProject</td>
<td><a href="https://www.ncbi.nlm.nih.gov/bioproject/PRJNA531376">PRJNA531376</a></td>
</tr>
<tr>
<td>BioSample</td>
<td><a href="https://www.ncbi.nlm.nih.gov/biosample/SAMN11357168">SAMN11357168</a></td>
</tr>
</table>
