<h2>Description</h2>
<p>This track shows peptide to spectra maps (PSMs) from mass spectrometry
data.</p>

<p>The Id of each track can be interpreted by splitting the string on each pipe
symbol "|". The fields are:</p>
<ol>
	<li>Annotation type</li>
	<li>Search data base</li>
	<li>Conitg/ Plasmid Id</li>
	<li>Experiment name</li>
	<li>Scan number _ match number</li>
</ol>

<h2>Display Conventions and Configuration</h2>
<p>The colours of the mass spectrometry data corresponds with the reading frame of the PSM:</p>
<ul>
<li><span style='border: 2px inset #000000; background-color:#ff1a1a;'>&nbsp;Red&nbsp;</span> + strand 1 frame
<li><span style='border: 2px inset #000000; background-color:#1a1aff;'>&nbsp;Blue&nbsp;</span> + strand 2 frame
<li><span style='border: 2px inset #000000; background-color:#2db92d;'>&nbsp;Green&nbsp;</span> + strand 3 frame
<li><span style='border: 2px inset #000000; background-color:#ac00e6;'>&nbsp;Purple&nbsp;</span> - strand 4 frame
<li><span style='border: 2px inset #000000; background-color:#e67300;'>&nbsp;Orange&nbsp;</span> - strand 5 frame
<li><span style='border: 2px inset #000000; background-color:#00e6b8;'>&nbsp;Turquoise&nbsp;</span> - strand 6 frame
</ul>

<h2>Methods</h2>
<p> The samples derive from a mixed culture of 8 bacteria (SIHUMIx)
Anaerostipes caccae (DSMZ 14662), Bacteroides thetaiotaomicron (DSMZ 2079),
Bifidobacterium longum (NCC 2705), Blautia producta (DSMZ 2950), Clostridium
butyricum (DSMZ 10702), Clostridium ramosum (DSMZ 1402), Escherichia coli K-12
(MG1655) and Lactobacillus plantarum (DSMZ 20174). The data set comprise 166
independent measurements, of which 90 used a standard protein preparation
protocol and the remaining 76 cover different enrichment protocols to elevate
the level of small proteins in solution.</p>

<p>Comet (2019.01 rev. 4) was used to generated PSMs. The search algorithm ran
against a 6frame data base of the whole genome which was generated by getorf of
the EMBOSS suite. To generate bed files from the resulting output a script was
used which is available under
<a href="">github</a>.</p>

<h2>Credits</h2>
<p>
Data were processed at the Bioinformatic Institute of Leipzig. For inquiries, please
contact us at the following address: john (at) bioinf.uni-leipzig.de
</p>

<h2>References</h2>
<p>The full details of the sample preparation can be found in <a
href="http://dx.doi.org/10.1016/j.jprot.2019.103604">Petruschke et al.</a> and
<a href="http://europepmc.org/article/MED/31816881">Sch&#228;pe et al.</a>.</p>
<p>The full data set is publicly available via ProteomeXchange with identifier
PXD023243.</p>
